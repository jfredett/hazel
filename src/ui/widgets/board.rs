#![allow(unused_imports, dead_code)]

use ratatui::prelude::*;
use crate::board::simple::PieceBoard;
use crate::board::Query;
use crate::types::{self, Occupant, Piece};
use crate::notation::*;

use ratatui::widgets::{Table, Row};

/// A widget what representeth a board in a 8x8 grid of 3x2 character cells. This is messy and bad
/// and will undergo some kind of finishing someday when I tire of it. Be ye warned, dragons doth
/// lie here.
#[derive(Default)]
pub struct Board<'a> {
    state: PieceBoard,
    board: Table<'a>
}

impl From<PieceBoard> for Board<'_> {
    fn from(state: PieceBoard) -> Self {
        let white_bg = Style::default().bg(WHITE_SQUARE).fg(Color::Black);
        let black_bg = Style::default().bg(BLACK_SQUARE).fg(Color::White);
        let white_triple = [white_bg].repeat(3);
        let black_triple = [black_bg].repeat(3);
        let white_first_row = Row::new([white_triple.clone(), black_triple.clone(), white_triple.clone(), black_triple.clone(), white_triple.clone(), black_triple.clone(), white_triple.clone(), black_triple.clone()].concat().into_iter().map(|style| Span::styled(" ", style)));
        let black_first_row = Row::new([black_triple.clone(), white_triple.clone(), black_triple.clone(), white_triple.clone(), black_triple.clone(), white_triple.clone(), black_triple.clone(), white_triple.clone()].concat().into_iter().map(|style| Span::styled(" ", style)));

        let table = Table::new([
            white_first_row.clone(),
            white_first_row.clone(),
            black_first_row.clone(),
            black_first_row.clone(),
            white_first_row.clone(),
            white_first_row.clone(),
            black_first_row.clone(),
            black_first_row.clone(),
            white_first_row.clone(),
            white_first_row.clone(),
            black_first_row.clone(),
            black_first_row.clone(),
            white_first_row.clone(),
            white_first_row.clone(),
            black_first_row.clone(),
            black_first_row.clone(),
        ], Constraint::from_maxes([1].repeat(24))).column_spacing(0);

        Self {
            state,
            board: table
        }
    }
}

/*
const ROOK: &'static str = "♜";
const KNIGHT : &'static str = "♞";
const BISHOP : &'static str = "♝";
const QUEEN : &'static str = "♛";
const KING : &'static str = "♚";
const PAWN : &'static str = "♟";
*/
const ROOK: &str = "R";
const KNIGHT : &str = "N";
const BISHOP : &str = "B";
const QUEEN : &str = "Q";
const KING : &str = "K";
const PAWN : &str = "P";


const WHITE_SQUARE : ratatui::prelude::Color = Color::Rgb(0x80, 0x80, 0x80);
const BLACK_SQUARE : ratatui::prelude::Color = Color::Rgb(0x40, 0x40, 0x40);

const BLACK_PIECE_WHITE_SQUARE : ratatui::prelude::Color = Color::Rgb(0x00, 0x00, 0x00);
const BLACK_PIECE_BLACK_SQUARE : ratatui::prelude::Color = Color::Rgb(0x00, 0x00, 0x00);
const WHITE_PIECE_WHITE_SQUARE : ratatui::prelude::Color = Color::Rgb(0xFF, 0xFF, 0xFF);
const WHITE_PIECE_BLACK_SQUARE : ratatui::prelude::Color = Color::Rgb(0xFF, 0xFF, 0xFF);

const MARK_COLOR : ratatui::prelude::Color = Color::Rgb(0x65, 0x65, 0x65);

impl Widget for &Board<'_> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        ratatui::prelude::Widget::render(&self.board, area, buf);

        let mut cursor = Square::by_rank_and_file();
        cursor.downward();
        for s in cursor {
            // Ratatui uses an Origin in the Top Left, with the first component being the
            // left/right offset ('file') and the second being the up/down offset ('rank').
            //
            // Chess, and in particular Hazel, uses an Origin in the Bottom Left, with the first
            // component being the rank and the second being the file. This is deeply annoying, but
            // can be neatly solved by doing some magic here.
            let buf_file : u16 = 14 - 2*s.rank() as u16;
            let buf_rank : u16 = 3*s.file() as u16 + 1;

            let cell = buf.get_mut(area.x + buf_rank, area.y + buf_file);
            let occ = self.state.get(s);

            match occ {
                Occupant::Occupied(piece, color) => {
                    let symbol = match piece {
                        Piece::Rook => { ROOK },
                        Piece::Knight => { KNIGHT },
                        Piece::Bishop => { BISHOP },
                        Piece::Queen => { QUEEN },
                        Piece::King => { KING },
                        Piece::Pawn => { PAWN }
                    };
                    cell.set_symbol(symbol);

                    match (cell.bg, color) {
                        (WHITE_SQUARE, types::Color::BLACK) => cell.fg = BLACK_PIECE_WHITE_SQUARE,
                        (WHITE_SQUARE, types::Color::WHITE) => cell.fg = WHITE_PIECE_WHITE_SQUARE,
                        (BLACK_SQUARE, types::Color::BLACK) => cell.fg = BLACK_PIECE_BLACK_SQUARE,
                        (BLACK_SQUARE, types::Color::WHITE) => cell.fg = WHITE_PIECE_BLACK_SQUARE,
                        _ => {}
                    }

                    cell.set_style(cell.style().bold());
                }
                Occupant::Empty => {} // Empty is the default state of the board.
            }

            // while we're here, I want to do write column/row information on the bottom bit of
            // the board in a nearly invisible color.
            let col_mark = buf.get_mut(area.x + buf_rank - 1, area.y + buf_file + 1);
            col_mark.set_char((b'a' + s.file() as u8) as char);
            col_mark.fg = MARK_COLOR;

            let row_mark = buf.get_mut(area.x + buf_rank, area.y + buf_file + 1);
            // note this sign flip, counting up from 'a' above, down from '8' here.
            row_mark.set_char((b'1' + s.rank() as u8) as char);
            row_mark.fg = MARK_COLOR;
        }


    }
}

#[cfg(test)]
mod tests {
    use ratatui::widgets::Cell;

    use super::*;

    #[test]
    fn renders_empty_board() {
        let rect = Rect::new(0, 0, 33, 17);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let board = PieceBoard::default();
        let board_widget = &Board::from(board);
        board_widget.render(rect, &mut buffer);

        let mut expected = Buffer::with_lines(vec![
            "                                 ",
            "a8 b8 c8 d8 e8 f8 g8 h8          ",
            "                                 ",
            "a7 b7 c7 d7 e7 f7 g7 h7          ",
            "                                 ",
            "a6 b6 c6 d6 e6 f6 g6 h6          ",
            "                                 ",
            "a5 b5 c5 d5 e5 f5 g5 h5          ",
            "                                 ",
            "a4 b4 c4 d4 e4 f4 g4 h4          ",
            "                                 ",
            "a3 b3 c3 d3 e3 f3 g3 h3          ",
            "                                 ",
            "a2 b2 c2 d2 e2 f2 g2 h2          ",
            "                                 ",
            "a1 b1 c1 d1 e1 f1 g1 h1          ",
            "                                 ",
        ]);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));


        assert_eq!(buffer, expected);
    }
}
