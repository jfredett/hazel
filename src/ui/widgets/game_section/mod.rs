pub mod board_section;
pub mod info_section;

pub struct GameSectionLayout {
    info_section: (),
    board_section: (),
}

