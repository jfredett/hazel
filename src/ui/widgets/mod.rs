pub mod hazel;
pub mod ply;
