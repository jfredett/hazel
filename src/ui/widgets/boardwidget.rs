use ratatui::prelude::*;
use ratatui::widgets::{Widget, Block, Borders};

use ratatui::buffer::Buffer;

use crate::ui::model::pieceboard::PieceBoard;

pub struct BoardWidget {
    board: PieceBoard
}

impl BoardWidget {
    // TODO: This should be an actual From implementation, so I can build these from multiple
    // sources (e.g., Ply)
    pub fn from(board: &PieceBoard) -> Self {
        Self {
            board: board.clone()
        }
    }
}


fn eight_cells(direction: Direction) -> Layout {
    Layout::default()
        .direction(direction)
        .constraints(
            [
                Constraint::Max(1),
                Constraint::Max(1),
                Constraint::Max(1),
                Constraint::Max(1),
                Constraint::Max(1),
                Constraint::Max(1),
                Constraint::Max(1),
                Constraint::Max(1),
                /* TODO: Would be cool to scale up the board dynamically with the allotted size.
                *  would need some scalable content for the cells.
                Constraint::Ratio(1, 8),
                Constraint::Ratio(1, 8),
                Constraint::Ratio(1, 8),
                Constraint::Ratio(1, 8),
                Constraint::Ratio(1, 8),
                Constraint::Ratio(1, 8),
                Constraint::Ratio(1, 8),
                Constraint::Ratio(1, 8),
                */
            ].as_ref()
        )
}

impl Widget for &BoardWidget {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let cols = eight_cells(Direction::Horizontal).split(area);
        let rows = [
            eight_cells(Direction::Vertical).split(cols[0]),
            eight_cells(Direction::Vertical).split(cols[1]),
            eight_cells(Direction::Vertical).split(cols[2]),
            eight_cells(Direction::Vertical).split(cols[3]),
            eight_cells(Direction::Vertical).split(cols[4]),
            eight_cells(Direction::Vertical).split(cols[5]),
            eight_cells(Direction::Vertical).split(cols[6]),
            eight_cells(Direction::Vertical).split(cols[7])
        ];

        for i in 0..8 {
            for j in 0..8 {
                let cell = Block::default()
                    .title(self.board.get(i,j).to_string())
                    .borders(Borders::NONE);
                cell.render(rows[j][i], buf);
            }
        }
    }
}
