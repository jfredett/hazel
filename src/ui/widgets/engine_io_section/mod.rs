pub mod outputwidget;

use ratatui::layout::Direction;
use ratatui::prelude::*;
use ratatui::widgets::{Block, Borders};
use crate::ui::app::Hazel;

// TODO: minimize the state by making EngineIOSection not need all of Hazel, instead it'll
// be it's own state management ViewModel and Hazel will construct it from itself.

pub struct EngineIOSection {
    output: Vec<String>,
    input: String,
}


struct Input {
}

struct Output {
}

impl Default for Input {
    fn default() -> Self {
        Self {}
    }
}

impl Default for Output {
    fn default() -> Self {
        Self {}
    }
}

impl StatefulWidget for &Input {
    type State = String;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let block = Block::default()
            .borders(Borders::NONE)
            .title(format!("$> {}", state))
            .border_style(Style::default().fg(Color::White).bg(Color::Black));
        block.render(area, buf);
    }
}

impl StatefulWidget for &Output {
    type State = Vec<String>;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let block = Block::default()
            .borders(Borders::ALL)
            .border_style(Style::default().fg(Color::White).bg(Color::Black));
        block.render(area, buf);

        // long lines are wrapped and indented by 2 spaces
        let output = state.clone();
        let mut adjusted = vec![];
        for line in output.iter() {
            if line.len() > area.width as usize - 2 {
                let mut working_copy = line.clone();
                let mut new_lines = vec![working_copy.split_off(area.width as usize - 2)];

                while working_copy.len() > area.width as usize - 2 {
                    let new_line = format!("  {}", working_copy.split_off(area.width as usize - 4));
                    new_lines.push(new_line);
                }
                new_lines.push(working_copy);
                adjusted.extend(new_lines.into_iter().rev());
            } else {
                adjusted.push(line.to_string());
            }
        }

        // should fill from the bottom to the top:
        let mut y = area.bottom() - 2;
        for line in adjusted.into_iter().rev() {
            buf.set_string(area.left() + 1, y, line, Style::default().fg(Color::White).bg(Color::Black));
            y -= 1;
        }
    }
}


lazy_static! {
    static ref LAYOUT : Layout = Layout::default()
        .direction(Direction::Vertical)
        .constraints(
            [
                Constraint::Fill(1),
                Constraint::Max(1),
            ]
                .as_ref(),
        );
}

impl EngineIOSection {
    pub fn new() -> Self {
        Self {
            output: vec![],
            input: String::new(),
        }
    }

    pub fn push(&mut self, line: String) {
        self.output.push(line);
    }
}

impl StatefulWidget for &EngineIOSection {
    type State = Hazel;
    fn render(self, area: Rect, buf: &mut Buffer, _state: &mut Self::State) {
        let chunks = LAYOUT.split(area);


        let output_widget = Output::default();
        output_widget.render(chunks[0], buf, &mut self.output.clone());

        let input_widget = Input::default();
        input_widget.render(chunks[1], buf, &mut self.input.clone());
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn placeholder() {
        let rect = Rect::new(0, 0, 64, 17);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let engine_io_section = &mut EngineIOSection::new();

        // Mock out the output from the 'engine'
        engine_io_section.push("Stockfish 16.1 by the Stockfish developers (see AUTHORS file)".to_string());
        engine_io_section.push("> isready".to_string());
        engine_io_section.push("readyok".to_string());
        engine_io_section.push("> position startpos moves".to_string());
        engine_io_section.push("> go depth 1".to_string());
        engine_io_section.push("info string NNUE evaluation using nn-baff1ede1f90.nnue".to_string());
        engine_io_section.push("info string NNUE evaluation using nn-b1a57edbea57.nnue".to_string());
        engine_io_section.push("info depth 1 seldepth 2 multipv 1 score cp 0 nodes 20 nps 1000 hashfull 0 tbhits 0 time 20 pv d2d4".to_string());
        engine_io_section.push("bestmove d2d4".to_string());

        engine_io_section.render(rect, &mut buffer, &mut Hazel::new());

        let mut expected = Buffer::with_lines(vec![
            "┌──────────────────────────────────────────────────────────────┐",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│Stockfish 16.1 by the Stockfish developers (see AUTHORS file) │",
            "│> isready                                                     │", // prefixes inputs by >
            "│readyok                                                       │",
            "│> position startpos moves                                     │",
            "│> go depth 1                                                  │",
            "│info string NNUE evaluation using nn-baff1ede1f90.nnue        │",
            "│info string NNUE evaluation using nn-b1a57edbea57.nnue        │",
            "│info depth 1 seldepth 2 multipv 1 score cp 0 nodes 20 nps 1000│", // wraps long lines?
            "│ hashfull 0 tbhits 0 time 20 pv d2d4                          │",
            "│bestmove d2d4                                                 │",
            "└──────────────────────────────────────────────────────────────┘",
            "$>                                                              ",
        ]);

        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));
        assert_eq!(buffer, expected);
    }
}


