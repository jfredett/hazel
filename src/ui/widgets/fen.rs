use ratatui::text::Text;
use ratatui::prelude::*;
use ratatui::widgets::Widget;

#[allow(unused_imports)]
use tracing::{debug, instrument};

use ratatui::buffer::Buffer;

use crate::ui::model::entry::Entry;
use crate::board::simple::PieceBoard;
use crate::board::interface::query;


#[derive(Debug, Default)]
pub struct FEN {
    // TODO: This can probably be aliased so I don't have to use it's full legal name here, but I
    // am lazy and don't want to read the docs.
    fen: crate::notation::fen::FEN,
    style: Style,
    alignment: Alignment
}

impl From<&Entry> for FEN {
    fn from(entry: &Entry) -> Self {
        Self::new(query::to_fen(&entry.boardstate))
    }
}

impl From<Entry> for FEN {
    fn from(entry: Entry) -> Self {
        Self::new(query::to_fen(&entry.boardstate))
    }
}

impl From<&PieceBoard> for FEN {
    fn from(board: &PieceBoard) -> Self {
        Self::new(query::to_fen(board))
    }
}

impl From<PieceBoard> for FEN {
    fn from(board: PieceBoard) -> Self {
        Self::new(query::to_fen(&board))
    }
}

impl FEN {
    pub fn new(fen: crate::notation::fen::FEN) -> Self {
        Self {
            fen,
            style: Style::default(),
            alignment: Alignment::Left
        }
    }

    pub fn set_style(mut self, style: Style) -> Self {
        self.style = style;
        self
    }

    pub fn alignment(mut self, alignment: Alignment) -> Self {
        self.alignment = alignment;
        self
    }

    #[inline(always)]
    pub fn center(self) -> Self {
        self.alignment(Alignment::Center)
    }
}

impl Widget for &FEN {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let fenstring = format!("{}", self.fen);
        Text::styled(fenstring, self.style).alignment(self.alignment).render(area, buf);
    }
}

#[cfg(test)]
mod tests {
    use crate::constants::START_POSITION_FEN;

    use super::*;

    #[test]
    fn renders_empty_fen_correctly() {
        let rect = Rect::new(0, 0, 64, 1);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));
        let board = PieceBoard::default();

        let fen_widget = &FEN::from(board);
        fen_widget.render(rect, &mut buffer);

        let mut expected = Buffer::with_lines(vec![
            "8/8/8/8/8/8/8/8 w KQkq - 0 1                                    "
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }

    #[test]
    fn render_startpos_fen_correctly() {
        let rect = Rect::new(0, 0, 64, 1);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let mut board = PieceBoard::default();
        board.set_fen(&crate::notation::fen::FEN::new(START_POSITION_FEN));

        let fen_widget = &FEN::from(board);
        fen_widget.render(rect, &mut buffer);

        let mut expected = Buffer::with_lines(vec![
            "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1        "
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);

    }
}
