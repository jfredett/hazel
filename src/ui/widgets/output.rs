
use ratatui::prelude::*;
use ratatui::widgets::{Block, Borders};

pub struct Output {
    buffer: Vec<String>
}

impl Default for Output {
    fn default() -> Self {
        Self {
            buffer: vec![]
        }
    }
}

impl Output {
    pub fn push(&mut self, line: String) {
        self.buffer.push(line);
    }

    pub fn buffer(&self) -> Vec<String> {
        self.buffer.clone()
    }
}


impl StatefulWidget for &Output {
    type State = Vec<String>;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let block = Block::default()
            .borders(Borders::ALL)
            .border_style(Style::default().fg(Color::White).bg(Color::Black));
        block.render(area, buf);

        // long lines are wrapped and indented by 2 spaces
        let output = state.clone();
        let mut adjusted = vec![];
        for line in output.iter() {
            if line.len() > area.width as usize - 2 {
                let mut working_copy = line.clone();
                let mut new_lines = vec![working_copy.split_off(area.width as usize - 2)];

                while working_copy.len() > area.width as usize - 2 {
                    let new_line = format!("  {}", working_copy.split_off(area.width as usize - 4));
                    new_lines.push(new_line);
                }
                new_lines.push(working_copy);
                adjusted.extend(new_lines.into_iter().rev());
            } else {
                adjusted.push(line.to_string());
            }
        }

        // should fill from the bottom to the top:
        let mut y = area.bottom() - 2;
        for line in adjusted.into_iter().rev() {
            buf.set_string(area.left() + 1, y, line, Style::default().fg(Color::White).bg(Color::Black));
            y -= 1;
        }
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn renders_as_expected() {
        let rect = Rect::new(0, 0, 64, 3);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let output = &mut Output::default();
        output.push("This is a test".to_string());
        output.render(rect, &mut buffer, &mut output.buffer().clone());

        let mut expected = Buffer::with_lines(vec![
            "┌──────────────────────────────────────────────────────────────┐",
            "│This is a test                                                │",
            "└──────────────────────────────────────────────────────────────┘",
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }

    #[test]
    fn renders_a_very_long_line_by_wrapping() {
        let rect = Rect::new(0, 0, 15, 6);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let output = &mut Output::default();
        output.push("This is a very long line that should be wrapped".to_string());
        output.render(rect, &mut buffer, &mut output.buffer().clone());

        let mut expected = Buffer::with_lines(vec![
        "┌─────────────┐",
        "│This is a ver│",
        "│y long line t│",
        "│hat should be│",
        "│ wrapped     │",
        "└─────────────┘",
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }
}







