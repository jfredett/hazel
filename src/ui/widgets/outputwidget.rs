use ratatui::prelude::*;
use ratatui::widgets::Widget;

use ratatui::buffer::Buffer;


#[derive(Debug, Default, Clone)]
pub struct OutputWidget(Vec<String>);

impl OutputWidget {
    pub fn empty() -> Self {
        Self {
            0: vec![]
        }
    }

    pub fn push(&mut self, message: String) {
        self.0.push(message);
    }
}

impl Widget for &OutputWidget {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let frame = Text::styled(
            self.0.join("\n"),
            Style::default().fg(Color::White).bg(Color::Black)
        );
        frame.render(area, buf);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn renders() {
        let rect = Rect::new(0, 0, 13, 1);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let output = &mut OutputWidget::empty();
        output.push("Hello, world!".to_string());
        output.render(buffer.area, &mut buffer);

        let mut expected = Buffer::with_lines(vec![
            "Hello, world!"
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }

    #[test]
    fn renders_multiline() {
        let rect = Rect::new(0, 0, 13, 3);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let output = &mut OutputWidget::empty();
        output.push("Hello, world!".to_string());
        output.push("Hello, world!".to_string());
        output.render(buffer.area, &mut buffer);

        let mut expected = Buffer::with_lines(vec![
            "Hello, world!",
            "Hello, world!",
            ""
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);

    }
}
