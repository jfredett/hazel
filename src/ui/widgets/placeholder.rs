use ratatui::widgets::{Block, Borders, Paragraph, StatefulWidget, Widget};
use ratatui::layout::{Alignment, Rect};
use ratatui::style::Style;
use ratatui::widgets::Wrap;

use ratatui::buffer::Buffer;


#[derive(Debug, Default, Clone)]
pub struct Placeholder {
    width: u16,
    height: u16,
    borders: Borders,
    text: &'static str,
    style: Style
}

// Features:
// 1. Renders to a specific size
// 2. mostly doesn't not work
// Planned Features:
// 1. tickers the text if size is too small. Maybe even DVD-logos in boxes?



/// A Placeholder widget that renders to the specific size given, with a border.
/// It contains the text "Placeholder" in the center of the widget.
impl Placeholder {
    pub fn of_size( width: u16, height: u16) -> Self {
        Self {
            width,
            height,
            borders: Borders::ALL,
            text: "Placeholder",
            style: Style::default()
        }
    }

    pub fn borders(mut self, borders: Borders) -> Self {
        self.borders = borders;
        self
    }

    pub fn text(mut self, text: &'static str) -> Self {
        self.text = text;
        self
    }

    fn calculate_text(&self) -> Paragraph {
        let text = [self.text].repeat(self.height as usize).join("\n");
        Paragraph::new(text)
            .style(self.style)
            .wrap(Wrap { trim: false })
    }

    pub fn set_style(mut self, style: Style) -> Self {
        self.style = style;
        self
    }
}

impl StatefulWidget for &Placeholder {
    type State = ();

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        Widget::render(self, area, buf);
    }
}

impl Widget for &Placeholder {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let widget = self.calculate_text().block(
            Block::default()
                .borders(self.borders)
        ).alignment(Alignment::Center);
        let new_rect = Rect::new(area.x, area.y, self.width, self.height);

        Widget::render(&widget, new_rect, buf);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ratatui::style::Color;

    #[test]
    fn calculates_text_correctly() {
        let placeholder = Placeholder::of_size(13, 3);
        let text = placeholder.calculate_text();

        let expected = Paragraph::new("Placeholder\nPlaceholder\nPlaceholder").wrap(Wrap { trim: false });
        assert_eq!(text, expected);
    }

    mod rendering {
        use super::*;

        #[test]
        fn smallest_box_without_overflow() {
            let rect = Rect::new(0, 0, 13, 3);
            let mut buffer = Buffer::empty(rect);
            buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            Widget::render(&Placeholder::of_size(13, 3), buffer.area, &mut buffer);

            let mut expected = Buffer::with_lines(vec![
                "┌───────────┐",
                "│Placeholder│",
                "└───────────┘"
            ]);
            expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            assert_eq!(buffer, expected);
        }

        #[test]
        fn mutliline_extrawide_box_odd_rows_even_columns() {
            let rect = Rect::new(0, 0, 20, 5);
            let mut buffer = Buffer::empty(rect);
            buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            Widget::render(&Placeholder::of_size(20, 5), buffer.area, &mut buffer);

            let mut expected = Buffer::with_lines(vec![
                "┌──────────────────┐",
                "│    Placeholder   │",
                "│    Placeholder   │",
                "│    Placeholder   │",
                "└──────────────────┘"
            ]);
            expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            assert_eq!(buffer, expected);
        }

        #[test]
        fn mutliline_extrawide_box_even_rows_odd_columns() {
            let rect = Rect::new(0, 0, 19, 6);
            let mut buffer = Buffer::empty(rect);
            buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            let placeholder = Placeholder::of_size(19, 6);
            Widget::render(&placeholder, buffer.area, &mut buffer);

            let mut expected = Buffer::with_lines(vec![
                "┌─────────────────┐",
                "│   Placeholder   │",
                "│   Placeholder   │",
                "│   Placeholder   │",
                "│   Placeholder   │",
                "└─────────────────┘"
            ]);
            expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            assert_eq!(buffer, expected);
        }

        #[test]
        fn mutliline_extrawide_box_even_rows_even_columns() {
            let rect = Rect::new(0, 0, 20, 6);
            let mut buffer = Buffer::empty(rect);
            buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            let placeholder = Placeholder::of_size(20, 6);
            Widget::render(&placeholder, buffer.area, &mut buffer);

            let mut expected = Buffer::with_lines(vec![
                "┌──────────────────┐",
                "│    Placeholder   │",
                "│    Placeholder   │",
                "│    Placeholder   │",
                "│    Placeholder   │",
                "└──────────────────┘"
            ]);
            expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            assert_eq!(buffer, expected);
        }

        #[test]
        fn mutliline_extrawide_box_odd_rows_odd_columns() {
            let rect = Rect::new(0, 0, 19, 5);
            let mut buffer = Buffer::empty(rect);
            buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            let placeholder = Placeholder::of_size(19, 5);
            Widget::render(&placeholder, buffer.area, &mut buffer);

            let mut expected = Buffer::with_lines(vec![
                "┌─────────────────┐",
                "│   Placeholder   │",
                "│   Placeholder   │",
                "│   Placeholder   │",
                "└─────────────────┘"
            ]);
            expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            assert_eq!(buffer, expected);
        }

        #[test]
        fn box_with_overflow() {
            let rect = Rect::new(0, 0, 10, 3);
            let mut buffer = Buffer::empty(rect);
            buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            let placeholder = Placeholder::of_size(10, 3);
            Widget::render(&placeholder, buffer.area, &mut buffer);

            let mut expected = Buffer::with_lines(vec![
                "┌────────┐",
                "│Placehol│",
                "└────────┘"
            ]);
            expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            assert_eq!(buffer, expected);
        }

        #[test]
        fn box_with_overflow_even_sides() {
            let rect = Rect::new(0, 0, 10, 4);
            let mut buffer = Buffer::empty(rect);
            buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            let placeholder = Placeholder::of_size(10, 4);
            Widget::render(&placeholder, buffer.area, &mut buffer);

            let mut expected = Buffer::with_lines(vec![
                "┌────────┐",
                "│Placehol│",
                "│   der  │",
                "└────────┘"
            ]);
            expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            assert_eq!(buffer, expected);
        }

        #[test]
        fn box_with_overflow_odd_sides() {
            let rect = Rect::new(0, 0, 11, 3);
            let mut buffer = Buffer::empty(rect);
            buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            let placeholder = Placeholder::of_size(11, 3);
            Widget::render(&placeholder, buffer.area, &mut buffer);

            let mut expected = Buffer::with_lines(vec![
                "┌─────────┐",
                "│Placehold│",
                "└─────────┘"
            ]);
            expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

            assert_eq!(buffer, expected);
        }
    }
}

