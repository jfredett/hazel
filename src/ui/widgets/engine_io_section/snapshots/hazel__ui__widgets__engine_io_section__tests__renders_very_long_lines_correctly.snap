---
source: src/ui/widgets/engine_io_section/mod.rs
expression: buffer
---
Buffer {
    area: Rect { x: 0, y: 0, width: 27, height: 6 },
    content: [
        "┌─────────────────────────┐",
        "│A line exceeding, the buf│",
        "│fers shortened width, sho│",
        "│uld be wrapped around.   │",
        "└─────────────────────────┘",
        "$>                         ",
    ],
    styles: [
        x: 0, y: 0, fg: White, bg: Black, underline: Reset, modifier: NONE,
    ]
}
